tests/std/tests/P2505R5_monadic_functions_for_std_expected/test.cpp

```diff
@@ -375,7 +375,153 @@ constexpr bool test() {
     return true;
 }
 
+template <class T, template <class...> class Tmpl>
+constexpr bool is_specialization_of = false;
+
+template <template <class...> class Tmpl, class... Args>
+constexpr bool is_specialization_of<Tmpl<Args...>, Tmpl> = true;
+
+template <class T>
+    requires is_specialization_of<remove_cvref_t<T>, expected>
+using expected_value_t = typename remove_cvref_t<T>::value_type;
+
+template <class T>
+    requires is_specialization_of<remove_cvref_t<T>, expected>
+using expected_error_t = typename remove_cvref_t<T>::error_type;
+
+template <class R>
+struct DefaultTransformer {
+    constexpr R operator()() const {
+        return R();
+    }
+
+    constexpr R operator()(auto&&) const {
+        return R();
+    }
+};
+
+template <class T>
+concept CanAndThen =
+    is_specialization_of<remove_cvref_t<T>, expected>
+    && requires(T&& t) { forward<T>(t).and_then(DefaultTransformer<expected<void, expected_error_t<T>>>{}); };
+
+template <class T>
+concept CanOrElse =
+    is_specialization_of<remove_cvref_t<T>, expected>
+    && requires(T&& t) { forward<T>(t).or_else(DefaultTransformer<expected<expected_value_t<T>, char>>{}); };
+
+template <class T>
+concept CanTransform = is_specialization_of<remove_cvref_t<T>, expected>
+                    && requires(T&& t) { forward<T>(t).transform(DefaultTransformer<expected_value_t<T>>{}); };
+
+template <class T>
+concept CanTransformError =
+    is_specialization_of<remove_cvref_t<T>, expected>
+    && requires(T&& t) { forward<T>(t).transform_error(DefaultTransformer<expected_error_t<T>>{}); };
+
+enum class HasMutCopy : bool { No, Yes };
+
+enum class HasConstCopy : bool { No, Yes };
+
+enum class HasMutMove : bool { No, Yes };
+
+enum class HasConstMove : bool { No, Yes };
+
+template <HasMutCopy MutCopyStatus, enum HasConstCopy ConstCopyStatus, HasMutMove MutMoveStatus,
+    HasConstMove ConstMoveStatus>
+struct State {
+    State() = default;
+    // clang-format off
+    State(State&) requires (static_cast<bool>(MutCopyStatus))           = default;
+    State(State&) requires (!static_cast<bool>(MutCopyStatus))          = delete;
+    State(const State&) requires (static_cast<bool>(ConstCopyStatus))   = default;
+    State(const State&) requires (!static_cast<bool>(ConstCopyStatus))  = delete;
+    State(State&&) requires (static_cast<bool>(MutMoveStatus))          = default;
+    State(State&&) requires (!static_cast<bool>(MutMoveStatus))         = delete;
+    State(const State&&) requires (!static_cast<bool>(ConstMoveStatus)) = delete;
+    // clang-format on
+
+    template <class U = State>
+        requires (!static_cast<bool>(ConstCopyStatus) && static_cast<bool>(ConstMoveStatus))
+    constexpr State(const type_identity_t<U>&&) noexcept {}

```

Comment:
No change requested: In the future, it would be good to explain subtleties like this in the PR description (if they aren't commented in the code). I figured this was intentional (versus just defaulting `State(const State&&)`) but had to try it to see:

```
D:\GitHub\STL\tests\std\tests\P2505R5_monadic_functions_for_std_expected\test.cpp(441): error C2610: 'State<MutCopyStatus,ConstCopyStatus,MutMoveStatus,ConstMoveStatus>::State(const State<MutCopyStatus,ConstCopyStatus,MutMoveStatus,ConstMoveStatus> &&)': is not a special member function or comparison operator which can be defaulted
D:\GitHub\STL\tests\std\tests\P2505R5_monadic_functions_for_std_expected\test.cpp(441): note: the argument must be a non-const rvalue reference
D:\GitHub\STL\tests\std\tests\P2505R5_monadic_functions_for_std_expected\test.cpp(447): note: see reference to class template instantiation 'State<MutCopyStatus,ConstCopyStatus,MutMoveStatus,ConstMoveStatus>' being compiled
```

I actually suspect that this is a compiler bug (I see *no* special-casing in the Standard, WG21-N4928 \[class.copy.ctor\]/2 says "A non-template constructor for class `X` is a move constructor if its first parameter is of type `X&&`, `const X&&`, `volatile X&&`, or `const volatile X&&`"), but I don't care enough about this squirrelly case to report it. :chipmunk: :joy_cat:

Reply:
`State(const State&&) = default;` would be ill-formed per https://eel.is/c++draft/dcl.fct.def.default since `State`'s implicitly-declared move constructor would have a parameter type of `State&&`.